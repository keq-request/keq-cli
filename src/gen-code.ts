import * as SwaggerParser from '@apidevtools/swagger-parser'
import * as chalk from 'chalk'
import * as changeCase from 'change-case'
import * as fs from 'fs-extra'
import { OpenAPIV3 } from 'openapi-types'
import * as pMap from 'p-map'
import * as path from 'path'
import * as R from 'ramda'
import * as semver from 'semver'
import { File } from './interface/file'
import { Options } from './interface/options'
import { getSafeOperationName } from './utils/get-safe-operation-name'
import { readTemplate } from './utils/read-template'

// import Handlebars from './handlebars-helper.js'
import * as Handlebars from 'handlebars'
import './handlebar/register-helper.js'
import './handlebar/register-partial.js'

const readAndCompileTemplate = (filename: string): HandlebarsTemplateDelegate => Handlebars.compile(readTemplate(filename))
const templates = {
  t_schema: readAndCompileTemplate('schema'),
  t_schema_exports: readAndCompileTemplate('schema-exports'),
  t_operation: readAndCompileTemplate('operation'),
  t_operation_exports: readAndCompileTemplate('operation-exports'),
}


export async function compile(moduleName: string, json: string | OpenAPIV3.Document, options?: Partial<Options>): Promise<File[]> {
  const swaggerParser = new SwaggerParser()
  let api: OpenAPIV3.Document
  try {
    await swaggerParser.bundle(json)

    if (!('openapi' in swaggerParser.api && semver.satisfies(swaggerParser.api.openapi, '^3'))) throw new Error('Only supports OpenAPI3')
    api = swaggerParser.api as OpenAPIV3.Document
  } catch (e) {
    console.warn(chalk.yellow('Swagger file does not conform to the swagger@3.0 standard specifications or have grammatical errors, which may cause unexpected errors'))
    api = json as any
  }

  const fileNamingStyle = options?.fileNamingStyle || 'snakeCase'
  const formatFilename = changeCase[fileNamingStyle]
  const outdir = options?.outdir || `${process.cwd()}/api`
  const output = path.join(outdir, formatFilename(moduleName))
  const requestInstance = options?.request ? path.relative(output, options.request) : 'keq'

  const files: File[] = []
  if (api.components?.schemas && !R.isEmpty(api.components.schemas)) {
    for (const [name, schema] of R.toPairs(api.components.schemas)) {
      const fileContent = templates.t_schema({
        api,
        name,
        schema,
        options: {
          fileNamingStyle,
        },
      })
      const filename = formatFilename(name)
      const filepath = path.join(output, 'components', 'schemas', `${filename}.ts`)

      files.push({
        name: filename,
        path: filepath,
        content: fileContent,
      })
    }

    const schemaExportsFilepath = path.join(output, 'components', 'schemas', 'index.ts')
    const schemaExportsFileContent = templates.t_schema_exports({
      api,
      options: {
        fileNamingStyle,
      },
    })

    files.push({
      name: 'index.ts',
      path: schemaExportsFilepath,
      content: schemaExportsFileContent,
    })
  }

  if (api.paths && !R.isEmpty(api.paths)) {
    for (const [pathname, pathItem] of R.toPairs(api.paths)) {
      if (!pathItem) continue

      for (const [method, operation] of R.toPairs(pathItem)) {
        if (typeof operation === 'object' && !Array.isArray(operation)) {
          const fileContent = templates.t_operation({
            api,
            moduleName,
            pathname,
            method,
            operation,
            options: {
              fileNamingStyle,
              request: requestInstance,
            },
          })

          const filename = formatFilename(getSafeOperationName(pathname, method, operation))
          const filepath = path.join(output, `${filename}.ts`)

          files.push({
            name: filename,
            path: filepath,
            content: fileContent,
          })
        } else {
          console.warn(chalk.yellow(`Operation ${String(method)} on path ${String(pathname)} cannot compiled, skipping`))
        }
      }
    }

    const operationExportsFilepath = path.join(output, 'index.ts')
    const operationExportsFileContent = templates.t_operation_exports({
      api,
      options: {
        fileNamingStyle,
      },
    })

    files.push({
      name: 'index.ts',
      path: operationExportsFilepath,
      content: operationExportsFileContent,
    })
  }

  return files
}

export async function genCode(moduleName: string, json: string | OpenAPIV3.Document, options: Options): Promise<void> {
  const files = await compile(moduleName, json, options)

  await pMap(
    files,
    async file => {
      await fs.ensureFile(file.path)
      await fs.writeFile(file.path, file.content)
    },
    { concurrency: 10 },
  )
}
