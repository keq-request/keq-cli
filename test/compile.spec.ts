import { expect, test } from '@jest/globals'
import { OpenAPIV3 } from 'openapi-types'
import { compile } from '~/gen-code'
import * as swagger from './swagger.json'


test('compile swagger', async () => {
  const files = await compile('test', swagger as unknown as OpenAPIV3.Document, {
    outdir: './outdir',
  })

  expect(files).toMatchSnapshot()
})

