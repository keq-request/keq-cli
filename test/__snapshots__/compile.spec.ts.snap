// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile swagger 1`] = `
[
  {
    "content": "import type { ArrayOfStringsSchema } from "./array_of_strings_schema"
import type { ObjectWithSpecifiedTypeAdditionalPropertiesSchema } from "./object_with_specified_type_additional_properties_schema"
import type { ObjectWithAdditionalPropertiesSchema } from "./object_with_additional_properties_schema"


/**
 * @interface RootSchema
 * @export
 */
export interface RootSchema {
  "unknownProp"?: any
  "constProp": "constValue"
  "enumProp"?: "str" | 1 | {
    "key": true
  }
  /**
   * @deprecated
   */
  "deprecatedStringProp"?: string
  /**
   * Deprecated Number Prop
   *
   * @deprecated
   */
  "deprecatedNumberProp"?: string
  "emptyObjectProp"?: {
  }
  /**
   * This is String Prop
   */
  "stringProp"?: string
  /**
   * Multiple line description
   * This is Number Prop
   */
  "numberProp"?: number
  "booleanProp"?: boolean
  /**
   * @type int64
   */
  "integerProp"?: number
  "tupleProp"?: [string, number, boolean]
  "stringPropWithMaxLength"?: string
  "arrayOfStringsProp"?: (string)[]
  "arrayOfNullableRefProp"?: (ObjectWithSpecifiedTypeAdditionalPropertiesSchema | null)[]
  "refProp"?: ArrayOfStringsSchema
  "objectProp"?: {
    "stringPropRequired": string
    "stringPropNullable"?: string | null
    "refPropNullable"?: ObjectWithAdditionalPropertiesSchema | null
  }
}
",
    "name": "root_schema",
    "path": "outdir/test/components/schemas/root_schema.ts",
  },
  {
    "content": "/**
 * @interface ArrayOfStringsSchema
 * @export
 */
export type ArrayOfStringsSchema = (string)[]
",
    "name": "array_of_strings_schema",
    "path": "outdir/test/components/schemas/array_of_strings_schema.ts",
  },
  {
    "content": "/**
 * @interface ObjectWithAdditionalPropertiesSchema
 * @export
 */
export interface ObjectWithAdditionalPropertiesSchema {
  [key: string]: undefined | any
}
",
    "name": "object_with_additional_properties_schema",
    "path": "outdir/test/components/schemas/object_with_additional_properties_schema.ts",
  },
  {
    "content": "/**
 * @interface ObjectWithSpecifiedTypeAdditionalPropertiesSchema
 * @export
 */
export interface ObjectWithSpecifiedTypeAdditionalPropertiesSchema {
  /**
   * This is String Addition Prop
   */
  [key: string]: undefined | string
}
",
    "name": "object_with_specified_type_additional_properties_schema",
    "path": "outdir/test/components/schemas/object_with_specified_type_additional_properties_schema.ts",
  },
  {
    "content": "/**
 * @interface RefSchema
 * @export
 */
export type RefSchema = ArrayOfStringsSchema
",
    "name": "ref_schema",
    "path": "outdir/test/components/schemas/ref_schema.ts",
  },
  {
    "content": "/**
 * @interface NullableObjectSchema
 * @export
 */
export type NullableObjectSchema = {
  "stringProp"?: string
} | null
",
    "name": "nullable_object_schema",
    "path": "outdir/test/components/schemas/nullable_object_schema.ts",
  },
  {
    "content": "/**
 * @interface OneOfSchema
 * @export
 */
export type OneOfSchema = string | number | boolean
",
    "name": "one_of_schema",
    "path": "outdir/test/components/schemas/one_of_schema.ts",
  },
  {
    "content": "/**
 * @interface AnyOfSchema
 * @export
 */
export type AnyOfSchema = string | number | boolean
",
    "name": "any_of_schema",
    "path": "outdir/test/components/schemas/any_of_schema.ts",
  },
  {
    "content": "/**
 * @interface AllOfSchema
 * @export
 */
export type AllOfSchema = {
  "stringProp"?: string
} & {
  "numberProp"?: number
}
",
    "name": "all_of_schema",
    "path": "outdir/test/components/schemas/all_of_schema.ts",
  },
  {
    "content": "export * from "./root_schema"
export * from "./array_of_strings_schema"
export * from "./object_with_additional_properties_schema"
export * from "./object_with_specified_type_additional_properties_schema"
export * from "./ref_schema"
export * from "./nullable_object_schema"
export * from "./one_of_schema"
export * from "./any_of_schema"
export * from "./all_of_schema"
",
    "name": "index.ts",
    "path": "outdir/test/components/schemas/index.ts",
  },
  {
    "content": "import type { KeqOperation } from 'keq'


export interface ResponseMap {
  "200": unknown
}


export type QueryParameters = {
  "requiredQueryProp": string
  "optionalQueryProp"?: string
}

export type RouteParameters = {
  "id": string
}

export type HeaderParameters = {
  "headerProp": string
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "get",
    "path": "outdir/test/types/get.ts",
  },
  {
    "content": "import { Keq } from 'keq'
import { request } from 'keq'
import type { RequestParameters, ResponseMap, Operation } from "./types/get"



export function get<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.get<ResponseMap[STATUS]>("/:id")
    .option('module', {
      name: "test",
      pathname: "/:id",
    })

  const queryWrap = (value: any) => typeof value === 'boolean' ? String(value) : value

  if (arg && "requiredQueryProp" in arg) req.query("requiredQueryProp", queryWrap(arg["requiredQueryProp"]))
  if (arg && "optionalQueryProp" in arg) req.query("optionalQueryProp", queryWrap(arg["optionalQueryProp"]))
  if (arg && "headerProp" in arg) req.set("headerProp", arg["headerProp"])
  if (arg && "id" in arg) req.params("id", String(arg["id"]))

  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}
",
    "name": "get",
    "path": "outdir/test/get.ts",
  },
  {
    "content": "import type { KeqOperation } from 'keq'
import type { RootSchema } from "../components/schemas/root_schema"


export interface ResponseMap {
  "200": RootSchema
}


export type QueryParameters = {
}

export type RouteParameters = {
  "id": string
}

export type HeaderParameters = {
}

export type BodyParameters =(RootSchema)

export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "post_id",
    "path": "outdir/test/types/post_id.ts",
  },
  {
    "content": "import { Keq } from 'keq'
import { request } from 'keq'
import type { RequestParameters, ResponseMap, Operation } from "./types/post_id"



export function post__$id<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.post<ResponseMap[STATUS]>("/:id")
    .option('module', {
      name: "test",
      pathname: "/:id",
    })

  if (arg && "id" in arg) req.params("id", String(arg["id"]))
  if (arg && "unknownProp" in arg) req.send({ "unknownProp": arg["unknownProp"] })
  if (arg && "constProp" in arg) req.send({ "constProp": arg["constProp"] })
  if (arg && "enumProp" in arg) req.send({ "enumProp": arg["enumProp"] })
  if (arg && "deprecatedStringProp" in arg) req.send({ "deprecatedStringProp": arg["deprecatedStringProp"] })
  if (arg && "deprecatedNumberProp" in arg) req.send({ "deprecatedNumberProp": arg["deprecatedNumberProp"] })
  if (arg && "emptyObjectProp" in arg) req.send({ "emptyObjectProp": arg["emptyObjectProp"] })
  if (arg && "stringProp" in arg) req.send({ "stringProp": arg["stringProp"] })
  if (arg && "numberProp" in arg) req.send({ "numberProp": arg["numberProp"] })
  if (arg && "booleanProp" in arg) req.send({ "booleanProp": arg["booleanProp"] })
  if (arg && "integerProp" in arg) req.send({ "integerProp": arg["integerProp"] })
  if (arg && "tupleProp" in arg) req.send({ "tupleProp": arg["tupleProp"] })
  if (arg && "stringPropWithMaxLength" in arg) req.send({ "stringPropWithMaxLength": arg["stringPropWithMaxLength"] })
  if (arg && "arrayOfStringsProp" in arg) req.send({ "arrayOfStringsProp": arg["arrayOfStringsProp"] })
  if (arg && "arrayOfNullableRefProp" in arg) req.send({ "arrayOfNullableRefProp": arg["arrayOfNullableRefProp"] })
  if (arg && "refProp" in arg) req.send({ "refProp": arg["refProp"] })
  if (arg && "objectProp" in arg) req.send({ "objectProp": arg["objectProp"] })

  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}
",
    "name": "post_id",
    "path": "outdir/test/post_id.ts",
  },
  {
    "content": "import type { KeqOperation } from 'keq'


export interface ResponseMap {
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "get_simple",
    "path": "outdir/test/types/get_simple.ts",
  },
  {
    "content": "import { Keq } from 'keq'
import { request } from 'keq'
import type { RequestParameters, ResponseMap, Operation } from "./types/get_simple"



export function get__simple<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.get<ResponseMap[STATUS]>("/simple")
    .option('module', {
      name: "test",
      pathname: "/simple",
    })


  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}
",
    "name": "get_simple",
    "path": "outdir/test/get_simple.ts",
  },
  {
    "content": "import type { KeqOperation } from 'keq'


export interface ResponseMap {
  "200": {
}
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "post_simple",
    "path": "outdir/test/types/post_simple.ts",
  },
  {
    "content": "import { Keq } from 'keq'
import { request } from 'keq'
import type { RequestParameters, ResponseMap, Operation } from "./types/post_simple"



export function post__simple<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.post<ResponseMap[STATUS]>("/simple")
    .option('module', {
      name: "test",
      pathname: "/simple",
    })


  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}
",
    "name": "post_simple",
    "path": "outdir/test/post_simple.ts",
  },
  {
    "content": "import type { KeqOperation } from 'keq'


export interface ResponseMap {
}


export type QueryParameters = {
}

export type RouteParameters = {
}

export type HeaderParameters = {
}

export type BodyParameters ={}
export type RequestParameters = QueryParameters & RouteParameters & HeaderParameters & BodyParameters


export interface Operation<STATUS extends keyof ResponseMap> extends KeqOperation {
  requestParams: RouteParameters
  requestQuery: QueryParameters
  requestHeaders: HeaderParameters
  requestBody: BodyParameters
  responseBody: ResponseMap[STATUS]
}
",
    "name": "delete_keyword",
    "path": "outdir/test/types/delete_keyword.ts",
  },
  {
    "content": "import { Keq } from 'keq'
import { request } from 'keq'
import type { RequestParameters, ResponseMap, Operation } from "./types/delete_keyword"



export function deleteKeyword<STATUS extends keyof ResponseMap>(arg?: RequestParameters): Keq<ResponseMap[STATUS], Operation<STATUS>> {
  const req = request.delete<ResponseMap[STATUS]>("/keyword")
    .option('module', {
      name: "test",
      pathname: "/keyword",
    })


  return req as unknown as Keq<ResponseMap[STATUS], Operation<STATUS>>
}
",
    "name": "delete_keyword",
    "path": "outdir/test/delete_keyword.ts",
  },
  {
    "content": "export * from "./get"
export * from "./post_id"
export * from "./get_simple"
export * from "./post_simple"
export * from "./delete_keyword"
",
    "name": "index.ts",
    "path": "outdir/test/index.ts",
  },
]
`;
